arch/arm64/src/common/arm64_arch_timer.c

```diff
@@ -0,0 +1,245 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <debug.h>
+#include <assert.h>
+
+#include <nuttx/arch.h>
+#include <arch/irq.h>
+#include <arch/chip/chip.h>
+#include <nuttx/spinlock.h>
+
+#include "arm64_arch.h"
+#include "arm64_internal.h"
+#include "arm64_gic.h"
+#include "arm64_arch_timer.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+#define MIN_DELAY  (1000)
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static uint64_t     last_cycle;
+static uint64_t     cycle_per_tick;
+static uint32_t     arch_timer_rate;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+static inline void arm64_arch_timer_set_compare(uint64_t value)
+{
+  write_sysreg(value, cntv_cval_el0);
+}
+
+static inline void arm64_arch_timer_enable(unsigned char enable)
+{
+  uint64_t value;
+
+  value = read_sysreg(cntv_ctl_el0);
+
+  if (enable)
+    {
+      value |= CNTV_CTL_ENABLE_BIT;
+    }
+  else
+    {
+      value &= ~CNTV_CTL_ENABLE_BIT;
+    }
+
+  write_sysreg(value, cntv_ctl_el0);
+}
+
+static inline void arm64_arch_timer_set_irq_mask(bool mask)
+{
+  uint64_t value;
+
+  value = read_sysreg(cntv_ctl_el0);
+
+  if (mask)
+    {
+      value |= CNTV_CTL_IMASK_BIT;
+    }
+  else
+    {
+      value &= ~CNTV_CTL_IMASK_BIT;
+    }
+
+  write_sysreg(value, cntv_ctl_el0);
+}
+
+static inline uint64_t arm64_arch_timer_count(void)
+{
+  return read_sysreg(cntvct_el0);
+}
+
+static inline uint32_t arm64_arch_timer_get_cntfrq(void)
+{
+  return read_sysreg(cntfrq_el0);
+}
+
+#ifdef CONFIG_SCHED_TICKLESS
+static void arm64_arch_timer_compare_isr(int irq, uint64_t *regs, void *arg)
+{
+  irqstate_t    flags;
+  uint64_t      curr_cycle;
+  uint32_t      delta_ticks;
+
+  flags = spin_lock_irqsave(&g_arch_timer_lock);
+
+  curr_cycle    = arm64_arch_timer_count();
+  delta_ticks   = (uint32_t)((curr_cycle - last_cycle) / cycle_per_tick);
+
+  last_cycle += delta_ticks * cycle_per_tick;
+
+  arm_arch_timer_set_irq_mask(true);
+
+  spin_unlock_irqrestore(&g_arch_timer_lock, flags);
+
+  nxsched_process_timer();
+}
+
+#else
+
+static void arm64_arch_timer_compare_isr(int irq, uint64_t *regs, void *arg)
```

Comment:
Same as tickless version of this function:

Should be:

`static int arm64_arch_timer_compare_isr(int irq, FAR void *context, FAR void *arg)`

At top of function, add:

`regs = (uint64_t) context;`

At end of function, add:

`return OK;`

Reply:
@xiaoxiang781216 thank you. I will edit my comment above to remove FAR.